node/subsystem/src/util.rs

```diff
@@ -0,0 +1,563 @@
+pub fn signing_key(validators: &[ValidatorId], keystore: &KeyStorePtr) -> Option<ValidatorPair> {
+	let keystore = keystore.read();
+	validators
+		.iter()
+		.find_map(|v| keystore.key_pair::<ValidatorPair>(&v).ok())
+}
+
+/// Local validator information
+///
+/// It can be created if the local node is a validator in the context of a particular
+/// relay chain block.
+pub struct Validator {
+	signing_context: SigningContext,
+	key: ValidatorPair,
+	index: ValidatorIndex,
+}
+
+impl Validator {
+	/// Get a struct representing this node's validator if this node is in fact a validator in the context of the given block.
+	pub async fn new<SenderMessage>(
+		parent: Hash,
+		keystore: KeyStorePtr,
+		mut sender: mpsc::Sender<SenderMessage>,
+	) -> Result<Self, Error>
+	where
+		SenderMessage: TryFrom<AllMessages>,
+		<SenderMessage as TryFrom<AllMessages>>::Error: std::fmt::Debug,
+	{
+		// Note: request_validators and request_signing_context do not and cannot run concurrently: they both
+		// have a mutable handle to the same sender.
+		// However, each of them returns a oneshot::Receiver, and those are resolved concurrently.
+		let (validators, signing_context) = futures::try_join!(
+			request_validators(parent, &mut sender).await?,
+			request_signing_context(parent, &mut sender).await?,
+		)?;
+
+		Self::construct(&validators, signing_context, keystore)
+	}
+
+	/// Construct a validator instance without performing runtime fetches.
+	///
+	/// This can be useful if external code also needs the same data.
+	pub fn construct(
+		validators: &[ValidatorId],
+		signing_context: SigningContext,
+		keystore: KeyStorePtr,
+	) -> Result<Self, Error> {
+		let key = signing_key(validators, &keystore).ok_or(Error::NotAValidator)?;
+		let index = validators
+			.iter()
+			.enumerate()
+			.find(|(_, k)| k == &&key.public())
+			.map(|(idx, _)| idx as ValidatorIndex)
+			.expect("signing_key would have already returned NotAValidator if the item we're searching for isn't in this list; qed");
+
+		Ok(Validator {
+			signing_context,
+			key,
+			index,
+		})
+	}
+
+	/// Get this validator's id.
+	pub fn id(&self) -> ValidatorId {
+		self.key.public()
+	}
+
+	/// Get this validator's local index.
+	pub fn index(&self) -> ValidatorIndex {
+		self.index
+	}
+
+	/// Get the current signing context.
+	pub fn signing_context(&self) -> &SigningContext {
+		&self.signing_context
+	}
+
+	/// Sign a payload with this validator
+	pub fn sign<Payload: EncodeAs<RealPayload>, RealPayload: Encode>(
+		&self,
+		payload: Payload,
+	) -> Signed<Payload, RealPayload> {
+		Signed::sign(payload, &self.signing_context, self.index, &self.key)
+	}
+
+	/// Validate the payload with this validator
+	pub fn check_payload<Payload: EncodeAs<RealPayload>, RealPayload: Encode>(
+		&self,
+		signed: Signed<Payload, RealPayload>,
+	) -> Result<(), ()> {
+		signed.check_signature(&self.signing_context, &self.id())
```

Comment:
This could also check that `signed.validator_index() == self.validator_index`

Reply:
Although this should already be checked before calling the function, and the docs should mention that.

Reply:
The use-case for this function isn't really clear though, since we don't usually need to check the signatures on things we signed ourselves.

Reply:
Honestly, I included it mainly for symmetry: if it's possible for a validator to sign something directly, then it makes sense to also offer a method to check the signature. 

I'll add both the check and the documentation.